rust/stub/src/pe_loader.rs

```diff
@@ -0,0 +1,197 @@
+                .sections
+                .iter()
+                .map(|section| {
+                    section
+                        .virtual_address
+                        .checked_add(section.virtual_size)
+                        .ok_or(Status::LOAD_ERROR)
+                })
+                .collect::<Result<Vec<u32>, uefi::Status>>()?;
+
+            let length = usize::try_from(section_lengths.into_iter().max().unwrap_or(0))
+                .map_err(to_load_error)?;
+
+            let base = boot_services.allocate_pages(
+                AllocateType::AnyPages,
+                MemoryType::LOADER_CODE,
+                bytes_to_pages(length),
+            )? as *mut u8;
+
+            unsafe {
+                core::ptr::write_bytes(base, 0, length);
+                core::slice::from_raw_parts_mut(base, length)
+            }
+        };
+
+        // Populate all sections in virtual memory.
+        for section in &pe.sections {
+            let copy_size =
+                usize::try_from(u32::min(section.virtual_size, section.size_of_raw_data))
+                    .map_err(to_load_error)?;
+            let raw_start = usize::try_from(section.pointer_to_raw_data).map_err(to_load_error)?;
+            let raw_end = raw_start.checked_add(copy_size).ok_or(Status::LOAD_ERROR)?;
+            let virt_start = usize::try_from(section.virtual_address).map_err(to_load_error)?;
+            let virt_end = virt_start
+                .checked_add(copy_size)
+                .ok_or(Status::LOAD_ERROR)?;
+
+            image[virt_start..virt_end].copy_from_slice(&file_data[raw_start..raw_end]);
+        }
+
+        // Apply image base relocations.
+        if let Some(reloc_table) = pe
+            .header
+            .optional_header
+            .and_then(|h| *h.data_directories.get_base_relocation_table())
+        {
+            let reloc_start =
+                usize::try_from(reloc_table.virtual_address).map_err(to_load_error)?;
+            let reloc_end = reloc_start
+                .checked_add(usize::try_from(reloc_table.size).map_err(to_load_error)?)
+                .ok_or(Status::LOAD_ERROR)?;
+
+            let mut relocations = &image[reloc_start..reloc_end];
+
+            while !relocations.is_empty() {
+                let size_of_block =
+                    usize::try_from(u32::from_ne_bytes(relocations[4..8].try_into().unwrap()))
+                        .map_err(to_load_error)?;
+
+                for i in 0..(size_of_block - 8) {
+                    let offset_field_entry =
+                        u16::from_ne_bytes(relocations[2 * i + 8..2 * i + 10].try_into().unwrap());
+                    match offset_field_entry >> 12 {
+                        // IMAGE_REL_BASED_ABSOLUTE
+                        0 => {}
+                        ty => todo!("Unhandled relocation type: {ty}"),
+                    }
+                }
+
+                relocations = &relocations[size_of_block..];
```

Comment:
This parsing code is pretty hard to understand. goblin already comes with [scroll](https://docs.rs/scroll/0.11.0/scroll/) as a dependency, which we could use here to clear this code up.

Given that we basically only check whether there are any relocations here at all, we could also leave it as is. In this case, I would ask you to leave a comment here about improving the parsing.

PS. I removed the panic on unsupported relocations and chose to return an easily recognizable error instead. This would at least leave the user in the boot menu being able to select another item to boot instead of crashing the system.

Reply:
I'm not too fond of that code either. Incidentally, it actually contains a bug (missing division by 2), and I wondered why it doesn't trigger a panic. Turns out while the Linux kernel has a `.reloc` section with one `IMAGE_REL_BASED_ABSOLUTE` (i.e. pointless) relocation, this is actually not linked in the header. I'm probably going to remove the relocation processing code entirely and just return `UNSUPPORTED` (or abuse `INCOMPATIBLE_VERSION` like you did) whenever encountering any linked relocation table at all.

Reply:
Yes, something like that would be a compromise for the time being.